### PR TITLE
Add more logs and make sysusers works in dnf --installroot

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by bluca** [comment]
Looks like some leaks are introduced:

```
305/891 test-sysusers                           FAIL     0.22 s (exit status 1)

--- command ---
12:15:00 /home/runner/work/systemd/systemd/build/test/test-sysusers.sh /home/runner/work/systemd/systemd/build/systemd-sysusers
--- stdout ---
*** Running /home/runner/work/systemd/systemd/test/test-sysusers/test-1.input
--- stderr ---
Creating group g1 with gid 111.
Creating group u1 with gid 222.
Creating user u1 (n/a) with uid 222 and gid 222.

=================================================================
==20456==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 50 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c7bc54bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7f5c7b018c01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7f5c7b018c01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7f5c7b07be0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7f5c7b07c08b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7f5c7aeefd26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x55aa17c360a3 in write_temporary_gshadow ../src/sysusers/sysusers.c:743
    #7 0x55aa17c360a3 in write_files ../src/sysusers/sysusers.c:822
    #8 0x55aa17c46a1d in run ../src/sysusers/sysusers.c:2101
    #9 0x55aa17c46a1d in main ../src/sysusers/sysusers.c:2108
    #10 0x7f5c79b140b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 49 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c7bc54bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7f5c7b018c01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7f5c7b018c01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7f5c7b07be0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7f5c7b07c08b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7f5c7aeefd26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x55aa17c37dbf in write_temporary_passwd ../src/sysusers/sysusers.c:390
    #7 0x55aa17c37dbf in write_files ../src/sysusers/sysusers.c:826
    #8 0x55aa17c46a1d in run ../src/sysusers/sysusers.c:2101
    #9 0x55aa17c46a1d in main ../src/sysusers/sysusers.c:2108
    #10 0x7f5c79b140b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 49 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c7bc54bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7f5c7b018c01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7f5c7b018c01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7f5c7b07be0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7f5c7b07c08b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7f5c7aeefd26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x55aa17c39979 in write_temporary_shadow ../src/sysusers/sysusers.c:509
    #7 0x55aa17c39979 in write_files ../src/sysusers/sysusers.c:830
    #8 0x55aa17c46a1d in run ../src/sysusers/sysusers.c:2101
    #9 0x55aa17c46a1d in main ../src/sysusers/sysusers.c:2108
    #10 0x7f5c79b140b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c7bc54bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7f5c7b018c01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7f5c7b018c01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7f5c7b07be0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7f5c7b07c08b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7f5c7aeefd26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x55aa17c34d09 in write_temporary_group ../src/sysusers/sysusers.c:642
    #7 0x55aa17c34d09 in write_files ../src/sysusers/sysusers.c:818
    #8 0x55aa17c46a1d in run ../src/sysusers/sysusers.c:2101
    #9 0x55aa17c46a1d in main ../src/sysusers/sysusers.c:2108
    #10 0x7f5c79b140b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: 196 byte(s) leaked in 4 allocation(s).
```

**Comment by bluca** [comment]
Still unhappy:

```
Direct leak of 50 byte(s) in 1 object(s) allocated from:
    #0 0x7ffb014dabc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7ffb0089ec01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7ffb0089ec01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7ffb00901e0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7ffb0090208b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7ffb00775d26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x5623e95510c2 in write_temporary_gshadow ../src/sysusers/sysusers.c:745
    #7 0x5623e95510c2 in write_files ../src/sysusers/sysusers.c:824
    #8 0x5623e9561a3c in run ../src/sysusers/sysusers.c:2103
    #9 0x5623e9561a3c in main ../src/sysusers/sysusers.c:2110
    #10 0x7ffaff39a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 49 byte(s) in 1 object(s) allocated from:
    #0 0x7ffb014dabc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7ffb0089ec01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7ffb0089ec01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7ffb00901e0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7ffb0090208b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7ffb00775d26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x5623e9552dde in write_temporary_passwd ../src/sysusers/sysusers.c:392
    #7 0x5623e9552dde in write_files ../src/sysusers/sysusers.c:828
    #8 0x5623e9561a3c in run ../src/sysusers/sysusers.c:2103
    #9 0x5623e9561a3c in main ../src/sysusers/sysusers.c:2110
    #10 0x7ffaff39a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 49 byte(s) in 1 object(s) allocated from:
    #0 0x7ffb014dabc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7ffb0089ec01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7ffb0089ec01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7ffb00901e0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7ffb0090208b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7ffb00775d26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x5623e9554998 in write_temporary_shadow ../src/sysusers/sysusers.c:511
    #7 0x5623e9554998 in write_files ../src/sysusers/sysusers.c:832
    #8 0x5623e9561a3c in run ../src/sysusers/sysusers.c:2103
    #9 0x5623e9561a3c in main ../src/sysusers/sysusers.c:2110
    #10 0x7ffaff39a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

Direct leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x7ffb014dabc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x7ffb0089ec01 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x7ffb0089ec01 in path_join_internal ../src/basic/path-util.c:579
    #3 0x7ffb00901e0f in tempfn_xxxxxx ../src/basic/tmpfile-util.c:128
    #4 0x7ffb0090208b in fopen_temporary ../src/basic/tmpfile-util.c:29
    #5 0x7ffb00775d26 in fopen_temporary_label ../src/shared/tmpfile-util-label.c:21
    #6 0x5623e954fd28 in write_temporary_group ../src/sysusers/sysusers.c:644
    #7 0x5623e954fd28 in write_files ../src/sysusers/sysusers.c:820
    #8 0x5623e9561a3c in run ../src/sysusers/sysusers.c:2103
    #9 0x5623e9561a3c in main ../src/sysusers/sysusers.c:2110
    #10 0x7ffaff39a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: 196 byte(s) leaked in 4 allocation(s).
```

**Comment by keszybz** [comment]
I was being an idiot. Should be fine now.

**Review by bluca — APPROVED** [review]
lgtm

File: src/basic/copy.c [hunk 1]
```diff
@@ -1243,7 +1243,7 @@ int copy_access(int fdf, int fdt) {
         return 0;
 }
 
-int copy_rights(int fdf, int fdt) {
+int copy_rights_with_fallback(int fdf, int fdt, const char *patht) {
```
Comment (yuwata) [inline src/basic/copy.c:1246]: `patht`??

File: src/basic/copy.h [hunk 2]
```diff
@@ -64,5 +64,8 @@ static inline int copy_bytes(int fdf, int fdt, uint64_t max_bytes, CopyFlags cop
 
 int copy_times(int fdf, int fdt, CopyFlags flags);
 int copy_access(int fdf, int fdt);
-int copy_rights(int fdf, int fdt);
+int copy_rights_with_fallback(int fdf, int fdt, const char *patht);
```
Comment (yuwata) [inline src/basic/copy.h:67]: Same here. `patht` -> `path`?

File: src/basic/copy.c (hunk 1 above)
Comment (keszybz) [reply src/basic/copy.c:1246]: I was matching the existing convetion used in that function: `fdt` → `patht`. If the second path was needed, it'd be `fdf` → `pathf`. I would like less cryptic names, maybe `fd_from`, `fd_to`, `path_from`, `path_to`, but then I guess the whole file should be changed.

Comment (yuwata) [reply src/basic/copy.c:1246]: Ah, I see. Thanks.

File: src/basic/fs-util.c [hunk 3]
```diff
@@ -263,8 +267,14 @@ int fchmod_and_chown(int fd, mode_t mode, uid_t uid, gid_t gid) {
...
                         r = fchmod_opath(fd, minimal & 07777);
-                        if (r < 0)
-                                return r;
+                        if (r < 0) {
+                                if (!path || r != -ENOSYS)
+                                        return r;
+
+                                /* Fallback path which doesn't use /proc/self/fd/. */
+                                if (chmod(path, minimal & 07777) < 0)
+                                        return -errno;
```
Comment (poettering) [inline src/basic/fs-util.c:276]: maybe debug log about this? more often than not doing this fallback looks like a setup problem

**Comment by poettering** [comment]
@socketpair when doing chmod/chown on an inode we want to make sure that noone can swap it out while we manipulate it. Because of that we open the inode with O_PATH, and then chmod/chown it via `/proc/self/fd/<fd>`. The O_PATH makes sure we actually modify the inode, and not whatever it might point to. But the chmod/chown only works if /proc is around then.

**Pull request merged**